### PR TITLE
Minified UMD detection

### PR DESCRIPTION
**Pull request by lewisl9029**
Following up on https://twitter.com/lewisl9029/status/1674890868661325829, here's an early version of the PR for detecting minified UMD exports in esm-cjs-lexer. :)

Unfortunately, a lot of the code in here ended up looking rather nasty. I'm still pretty new to Rust, so would love to get some tips from someone more experienced on how the code could be improved! Happy to iterate on the code until you're happy with it.

I've added a number of minimal test cases representing some of the minified UMD patterns I've encountered in the wild (some from webpack, some from rollup). I'm sure there will be more, will tackle them as they come up, feel free to ping me in issues if you see any new cases that needs to be covered going forward!

I've been testing it out in prod by publishing it to this fork: https://www.npmjs.com/package/@lewisl9029/esm-cjs-lexer

And it seems to be working well so far, albeit maybe a bit slower (though I haven't done any benchmarks yet, please do let me know if you notice any performance footguns in here as well!). Feel free to take it for a spin when you have the chance and share your findings. Cheers!

### CODE REVIEW

File: packages/esm-cjs-lexer/src/cjs.rs [hunk 1]
```diff
@@ -524,6 +524,140 @@ impl ExportsParser {
...
+          self.get_webpack4_exports(&expr, webpack_exports_sym, webpack_require_sym)
+        }
+      }
+      Expr::Call(call) => match webpack_require_sym {
+        Some(webpack_require_sym) => {
+          if let Some(Expr::Member(MemberExpr { obj, prop, .. })) = with_expr_callee(call) {
+            if let (Expr::Ident(Ident { sym: obj_sym, .. }), MemberProp::Ident(Ident { sym: prop_sym, .. })) =
+              (&**obj, &*prop)
+            {
+              if obj_sym.as_ref().eq(webpack_require_sym) && prop_sym.as_ref().eq("r") {
```
Comment (ije) [inline packages/esm-cjs-lexer/src/cjs.rs:546]: the prop `r` is stable? is it generated by minifier?

File: packages/esm-cjs-lexer/src/test.rs [hunk 2]
```diff
@@ -698,6 +698,338 @@ mod tests {
...
+    "#;
+    let swc = SWC::parse("index.cjs", source).expect("could not parse module");
+    let (exports, _) = swc
+      .parse_cjs_exports("production", true)
+      .expect("could not parse exports");
+    assert_eq!(exports.join(","), "__esModule,named2,named1,default");
+  }
+
+  #[test]
+  fn parse_cjs_exports_case_21_8() {
```
Comment (ije) [inline packages/esm-cjs-lexer/src/test.rs:991]: oh, like this tests ♥️

File: packages/esm-cjs-lexer/src/cjs.rs (hunk 1 above)
Comment (lewisl9029) [reply packages/esm-cjs-lexer/src/cjs.rs:546]: Yep, they seem to be explicitly named that way (probably to save space?). Here's what the unminified version looks like:

```js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else {
		var a = factory();
		for(var i in a) (typeof exports === 'object' ? exports : root)[i] = a[i];
	}
})(this, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "named", function() { return named; });
var named = 'named-export';
/* harmony default export */ __webpack_exports__["default"] = ('default-export');

/***/ })
/******/ ]);
});
```

**Comment by ije** [comment]
@lewisl9029 your tests works great for me and i'm going to merge this. thanks for contributing!

**Review by ije — APPROVED** [review]

**Pull request merged**